### PR TITLE
mds: don't mark dirty rstat on non-auth inode

### PR DESCRIPTION
**Pull request by david-z**
Fixes: #http://tracker.ceph.com/issues/20569

Signed-off-by: Zhi Zhang <zhangz.david@outlook.com>

### CODE REVIEW

File: src/mds/MDCache.cc [hunk 1]
```diff
@@ -2166,7 +2166,8 @@ void MDCache::predirty_journal_parents(MutationRef mut, EMetaBlob *blob,
     // rstat
     if (!primary_dn) {
       // don't update parent this pass
-    } else if (!linkunlink && !(pin->nestlock.can_wrlock(-1) &&
+    } else if (pin->is_auth() && !linkunlink && 
+				!(pin->nestlock.can_wrlock(-1) &&
 				pin->versionlock.can_wrlock())) {
```
Comment (ukernel) [inline src/mds/MDCache.cc]: this change does not seem correct.  CInode::mark_dirty_rstat() calls Locker::mark_updated_scatterlock(), which adds parent inode to updated_scatterlocks list. Later Locker::scatter_tick() triggers scatter-gather processes, which should clear the dirtyrstat flag.

Comment (ukernel) [inline src/mds/MDCache.cc]: Actually, it's a scatter-gather bug. dirtyrstat inode isn't processed properly when it and its parent inode belong to different auth mds. CInode::finish_scatter_update() should call CDir::assimilate_dirty_rstat_inodes() when inode is not auth and next lock state is MIX.

Comment (david-z) [inline src/mds/MDCache.cc]: @ukernel  Thanks for the review. I can get most of your points, but one question here: why next lock state should be MIX?

Comment (ukernel) [inline src/mds/MDCache.cc]: CDir's rstat can be updated only when nestlock is in MIX state.

**Comment by ukernel** [comment]
do you want to write a new patch? if not, I will close this PR

**Comment by david-z** [comment]
@ukernel  I am working on it. Your suggestion could work, but it will cause CDir::commit crashing at assert(want > committed_version);

I am still looking at this crashing.

**Comment by david-z** [comment]
@ukernel Will you submit another patch?

**Pull request closed without merging**